utils/s3/client.cc

```diff
@@ -1092,6 +1093,125 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece, as));
 }
 
+class client::chunked_download_source final : public seastar::data_source_impl {
+    struct claimed_buffer {
+        temporary_buffer<char> _buffer;
+        semaphore_units<> _claimed_memory;
+        s3_clock::duration _consumption_time;
+        claimed_buffer(temporary_buffer<char>&& buf, semaphore_units<>&& claimed_memory, s3_clock::duration consumption_time)
+            : _buffer(std::move(buf)), _claimed_memory(std::move(claimed_memory)), _consumption_time(consumption_time) {}
+    };
+    shared_ptr<client> _client;
+    sstring _object_name;
+    seastar::abort_source* _as;
+    range _range;
+    static constexpr size_t _max_buffers_size = 5_MiB;
+    static constexpr double _buffers_low_watermark = 0.5;
+    std::deque<claimed_buffer> _buffers;
+    size_t _buffers_size = 0;
+    bool _is_finished = false;
+    condition_variable _bg_fiber_cv;
+    condition_variable _get_cv;
+    future<> _filling_fiber = make_ready_future<>();
+
+    future<> make_filling_fiber() {
+        s3l.trace("Fiber starts cycle for object '{}'", _object_name);
+        while (!_is_finished) {
+            try {
+                if (_buffers_size >= _max_buffers_size * _buffers_low_watermark) {
+                    co_await _bg_fiber_cv.when([this] { return _buffers_size < _max_buffers_size * _buffers_low_watermark; });
+                }
+
+                if (_is_finished) {
+                    s3l.trace("Abandoning fiber for object '{}'", _object_name);
+                    co_return;
+                }
+
+                auto req = http::request::make("GET", _client->_host, _object_name);
+                range current_range {_range.off, _max_buffers_size - _buffers_size};
```

Comment:
This breaks the single-downloader optimization. In this, the file will still be downloaded in 5mb chunks, even if buffer-s are get()-ed in a timely manner